osu.Game/Database/ImportTask.cs

```diff
@@ -46,55 +46,37 @@ public ImportTask(Stream stream, string filename)
         /// </summary>
         public ArchiveReader GetReader()
         {
-            return Stream != null
-                ? getReaderFrom(Stream)
-                : getReaderFrom(Path);
-        }
+            if (Stream == null)
+            {
+                if (ZipUtils.IsZipArchive(Path))
+                    return new ZipArchiveReader(File.Open(Path, FileMode.Open, FileAccess.Read, FileShare.Read), System.IO.Path.GetFileName(Path));
+                if (Directory.Exists(Path))
+                    return new DirectoryArchiveReader(Path);
+                if (File.Exists(Path))
+                    return new SingleFileArchiveReader(Path);
 
-        /// <summary>
-        /// Deletes the file that is encapsulated by this <see cref="ImportTask"/>.
-        /// </summary>
-        public virtual void DeleteFile()
-        {
-            if (File.Exists(Path))
-                File.Delete(Path);
-        }
+                throw new InvalidFormatException($"{Path} is not a valid archive");
+            }
 
-        /// <summary>
-        /// Creates an <see cref="ArchiveReader"/> from a stream.
-        /// </summary>
-        /// <param name="stream">A seekable stream containing the archive content.</param>
-        /// <returns>A reader giving access to the archive's content.</returns>
-        private ArchiveReader getReaderFrom(Stream stream)
-        {
-            if (!(stream is MemoryStream memoryStream))
+            if (Stream is MemoryStream memoryStream)
             {
-                // This isn't used in any current path. May need to reconsider for performance reasons (ie. if we don't expect the incoming stream to be copied out).
-                memoryStream = new MemoryStream(stream.ReadAllBytesToArray());
-                stream.Dispose();
-            }
+                if (ZipUtils.IsZipArchive(memoryStream))
+                    return new ZipArchiveReader(memoryStream, Path);
 
-            if (ZipUtils.IsZipArchive(memoryStream))
-                return new ZipArchiveReader(memoryStream, Path);
+                return new MemoryStreamArchiveReader(memoryStream, Path);
```

Comment:
The tests are failing because this is not a correct transcription of the previous logic. The failing case is a `Stream` that is not a `MemoryStream`, but whose underlying data contains a ZIP archive.

- Previously: "not a memory stream" check converts stream to a memory stream, which then gets treated as ZIP archive and read as ZIP
- Now: "not a memory stream" check just presumes it's a single opaque file, causing failures in tests to read `.osu` file out of `.osz`s (as the `.osz` is treated as a single blob).

Obvious fix would be something like

```diff
diff --git a/osu.Game/Database/ImportTask.cs b/osu.Game/Database/ImportTask.cs
index 6d432aac25..d827ad47bc 100644
--- a/osu.Game/Database/ImportTask.cs
+++ b/osu.Game/Database/ImportTask.cs
@@ -4,6 +4,7 @@
 using System.IO;
 using osu.Framework.Extensions;
 using osu.Game.IO.Archives;
+using osu.Game.Rulesets.Difficulty.Skills;
 using osu.Game.Utils;
 using SharpCompress.Common;
 
@@ -58,16 +59,16 @@ public ArchiveReader GetReader()
                 throw new InvalidFormatException($"{Path} is not a valid archive");
             }
 
-            if (Stream is MemoryStream memoryStream)
+            if (Stream is not MemoryStream memoryStream)
             {
-                if (ZipUtils.IsZipArchive(memoryStream))
-                    return new ZipArchiveReader(memoryStream, Path);
-
-                return new MemoryStreamArchiveReader(memoryStream, Path);
+                // Path used primarily in tests (converting `ManifestResourceStream`s to `MemoryStream`s).
+                memoryStream = new MemoryStream(Stream.ReadAllBytesToArray());
             }
 
-            // This isn't used in any current path. May need to reconsider for performance reasons (ie. if we don't expect the incoming stream to be copied out).
-            return new ByteArrayArchiveReader(Stream.ReadAllBytesToArray(), Path);
+            if (ZipUtils.IsZipArchive(memoryStream))
+                return new ZipArchiveReader(memoryStream, Path);
+
+            return new MemoryStreamArchiveReader(memoryStream, Path);
         }
 
         /// <summary>

```

I'll also mention that previously the provided `stream` would be disposed if it wasn't a `MemoryStream`. I fully realise that was _super_ dodgy, but if we don't bring that back, we may begin leaking stream instances.